File: code/dart/lib/src/http/http_service_client.dart
import 'dart:convert';
import 'dart:developer' as dev;

import 'package:http/http.dart' as http;

import '../core/service_core.dart';
import 'http_exceptions.dart';

class HttpServiceClient implements ServiceClient {
  HttpServiceClient(this.config, {http.Client? client})
      : _client = client ?? http.Client();

  final ServiceClientConfig config;
  final http.Client _client;

  @override
  Future<ServiceResponse> send(ServiceRequest request) async {
    if (request.protocol != ServiceProtocol.http) {
      throw Exception('Unsupported protocol for HttpServiceClient');
    }

    final url = config.baseUrl.resolve(request.endpoint);
    final effectiveHeaders = <String, String>{
      'Content-Type': 'application/json',
      ...config.defaultHeaders,
      if (request.headers != null) ...request.headers!,
    };

    String? encodeBody(Object? body, {bool useEmptyObjectWhenNull = false}) {
      if (body == null) {
        return useEmptyObjectWhenNull ? jsonEncode(<String, dynamic>{}) : null;
      }
      if (body is String) return body;
      return jsonEncode(body);
    }

    Future<http.Response> doCall() {
      final method = request.method.toUpperCase();
      switch (method) {
        case 'GET':
          return _client
              .get(url, headers: effectiveHeaders)
              .timeout(config.timeout);
        case 'POST':
          return _client
              .post(
                url,
                headers: effectiveHeaders,
                body: encodeBody(request.body, useEmptyObjectWhenNull: true),
              )
              .timeout(config.timeout);
        case 'PATCH':
          return _client
              .patch(
                url,
                headers: effectiveHeaders,
                body: encodeBody(request.body),
              )
              .timeout(config.timeout);
        default:
          final req = http.Request(method, url)
            ..headers.addAll(effectiveHeaders);
          final encoded = encodeBody(request.body);
          if (encoded != null) {
            req.body = encoded;
          }
          return _client
              .send(req)
              .then(http.Response.fromStream)
              .timeout(config.timeout);
      }
    }

    try {
      final response = await doCall();

      if (response.statusCode >= 200 && response.statusCode < 300) {
        return _toServiceResponse(response);
      }

      dev.log('HTTP Error Response:');
      dev.log('  Status: ${response.statusCode}');
      dev.log('  Body: ${response.body}');

      Map<String, dynamic>? responseData;
      try {
        responseData = jsonDecode(response.body) as Map<String, dynamic>?;
      } catch (_) {
        // Response body is not valid JSON — leave as null
      }

      throw HttpClientException(
        statusCode: response.statusCode,
        message: request.errorMessage,
        response: responseData,
      );
    } catch (e) {
      if (e is HttpClientException) rethrow;

      dev.log('HTTP Client Error: $e');
      throw Exception('${request.errorMessage}: [Connection error] - $e');
    }
  }

  @override
  Future<void> close() async {
    _client.close();
  }

  ServiceResponse _toServiceResponse(http.Response response) {
    final rawBody = response.body;
    final data = rawBody.isNotEmpty ? jsonDecode(rawBody) : null;
    return ServiceResponse(
      statusCode: response.statusCode,
      headers: response.headers,
      rawBody: rawBody,
      data: data,
    );
  }
}

File: code/ts/src/http/http_service_client.ts
import {
  type ServiceClient,
  ServiceClientConfig,
  ServiceRequest,
  type ServiceResponse,
  type ServiceProtocol,
} from '../core/service_client';
import { HttpClientException } from './http_client_exception';

export class HttpServiceClient implements ServiceClient {
  private readonly config: ServiceClientConfig;

  constructor(config: ServiceClientConfig) {
    this.config = config;
  }

  async send(request: ServiceRequest): Promise<ServiceResponse> {
    if (request.protocol !== ('http' satisfies ServiceProtocol)) {
      throw new Error('Unsupported protocol for HttpServiceClient');
    }

    const url = new URL(request.endpoint, this.config.baseUrl);
    const effectiveHeaders: Record<string, string> = {
      'Content-Type': 'application/json',
      ...this.config.defaultHeaders,
      ...(request.headers ?? {}),
    };

    const encodedBody = this.encodeBody(
      request.body,
      request.method.toUpperCase() === 'POST',
    );

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url.toString(), {
        method: request.method.toUpperCase(),
        headers: effectiveHeaders,
        body: encodedBody,
        signal: controller.signal,
      });

      clearTimeout(timer);

      const rawBody = await response.text();

      if (response.status >= 200 && response.status < 300) {
        const data = rawBody.length > 0 ? JSON.parse(rawBody) : undefined;
        return {
          statusCode: response.status,
          headers: Object.fromEntries(response.headers.entries()),
          rawBody,
          data,
        };
      }

      let responseData: Record<string, unknown> | undefined;
      try {
        responseData = JSON.parse(rawBody) as Record<string, unknown>;
      } catch {
        // Response body is not valid JSON — leave as undefined
      }

      throw new HttpClientException({
        statusCode: response.status,
        message: request.errorMessage,
        response: responseData,
      });
    } catch (e) {
      clearTimeout(timer);

      if (e instanceof HttpClientException) throw e;

      throw new Error(`${request.errorMessage}: [Connection error] - ${e}`);
    }
  }

  async close(): Promise<void> {
    // fetch does not maintain persistent connections — no-op
  }

  /// Encodes the request body as JSON string.
  /// POST sends '{}' when body is null/undefined (B-04).
  /// If body is already a string, pass as-is (B-05).
  private encodeBody(body: unknown, isPost: boolean): string | undefined {
    if (body == null) {
      return isPost ? '{}' : undefined;
    }
    if (typeof body === 'string') return body;
    return JSON.stringify(body);
  }
}
